src/data/simple_dmatrix.h

```diff
@@ -69,23 +77,22 @@ class SimpleDMatrix : public DMatrix {
               std::max(inferred_num_columns,
                        static_cast<uint64_t>(element.column_idx + 1));
           if (!common::CheckNAN(element.value) && element.value != missing) {
-            builder.AddBudget(element.row_idx, tid);
+            builder.AddBudget(element.row_idx - num_lines_processed, tid);
           }
         }
       }
       builder.InitStorage();
 
       // Second pass over batch, placing elements in correct position
 #pragma omp parallel for schedule(static)
-      for (omp_ulong i = 0; i < static_cast<omp_ulong>(num_lines);
-        ++i) {  // NOLINT(*)
+      for (omp_ulong i = 0; i < static_cast<omp_ulong>(num_lines); ++i) {  // NOLINT(*)
         int tid = omp_get_thread_num();
         auto line = batch.GetLine(i);
         for (auto j = 0ull; j < line.Size(); j++) {
           auto element = line.GetElement(j);
           if (!common::CheckNAN(element.value) && element.value != missing) {
-            builder.Push(element.row_idx, Entry(element.column_idx, element.value),
-              tid);
+            builder.Push(element.row_idx - num_lines_processed,
```

Comment:
Again, num_lines_processed != num_rows_processed.

Reply:
i also wonder if the parallel builder was build to handle this type of use-cases. i see them typically used for building csc type matrices. i do not know the history here, and thus the question. are we retrofitting this to solve the needs of the adapter (to also build csr style matrices)?

Reply:
Yes we are retrofitting it, it just wasn't designed to process more than one batch. Let me produce another PR for comparison that modifies the group builder to more efficiently handle this case.